File: Matlab/PenZDAval.m
function [val_w, DVs, gamma, best_ind, val_score, classMeans] = PenZDAval(train, val,D, gmults, consttype, sparsity_level, beta, tol, maxits,quiet)
% PENZDAVAL performs validation to train regularization parameters and DVs.
%
% INPUT.
% train, val: training and validation set.
% D: Penalty dictionary basis matrix 
% gmults: list of multipliers of maximum gamma (calculated later) to compare.
% consttype: "ball/sphere" imposes ball/spherical constraint on DVs.
% sparsity_level: desired minimum sparsity level for validation scoring.
% beta: augmented Lagrangian penalty parameter.
% tol: stopping tolerances
% maxits: maximum number of iterations.
% quiet = true suppresses display of intermediate outputs.
% OUTPUT.
% val_w: discriminant vectors corresponding to best parameter.
% DVs: set of all discriminant vectors.
% gamma: best parameter choice.
% best_ind: position of best gamma.
% val_score: array of validation scores.
% classMeans: set of training data class means.

% Initialize classMeans of training data and get number of classes.
% classes=train(:,1);
[n,p]=size(train);
p = p-1;
% X=train(:,2:p);
% %X=normalize(X);
% %Extract observations 
% labels=unique(classes); 
% K=length(labels);
% %Initiate matrix of within-class means
% p=p-1;
% classMeans=zeros(p,K);
% R=zeros(K,p);
% 
% %for each class, make an object in the list containing only the obs of that
% %class and update the between and within-class sample
% M=zeros(p,n);
% 
% 
% % Calculate scaled classMeans for calculating R and N in objective function.
% for i=1:K    
%     class_obs=X(classes==labels(i),:);
%     %Get the number of obs in that class (the number of rows)
%     ni=size(class_obs,1);
%     %Compute within-class mean
%     classMeans(:,i)=mean(class_obs);
%     %Update W 
%     xj=class_obs-ones(ni,1)*classMeans(:,i)';
%     M(:,classes == labels(i)) =xj';
%     R(i,:)= sqrt(ni)*mean(class_obs)';
% end

% Call calcClassMeans to compute class-means and covariance matrices.
[classMeans, K, M, R]=calcClassMeans(train);

% Calculate null basis.
N=null(M);

%Compute leading eigenvector of N'*B*N
RN = R*N;
%size(RN)
[~,sigma,w]=svds(RN, 1,'largest');
% normalize R.
R=R/sigma;
RN = RN/sigma;
RN0 = RN;
Nw = N*w;

% Define d operators.
if isdiag(D)
    % Check if D = I
    d = diag(D); 
    if norm(d - ones(p,1)) < 1e-12 % D=I
        Dx = @(x) x;
        Dtx = Dx;
    else % D ~=I
        Dx = @(x) d.*x; % diagonal scaling if D is diagonal.
        Dtx = Dx;
    end
else
    Dx = @(x) D*x;
    Dtx = @(x) D'*x;
end

%++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
%Initialize the validation scores
%++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
num_gammas = length(gmults);
val_score= (p+1)*ones(num_gammas,1);
%mc_ind=1;
%l0_ind=1;
best_ind=1;
%min_mc=1;
%min_l0=p;
triv=0;

%Initialize DVs and iterations
N0=N;
DVs=zeros(p,K-1,num_gammas);
its=zeros(num_gammas,1);

%For each gamma, calculate ZVDs and corresponding validation score
gammas=zeros(num_gammas, K-1);
gmax = norm(RN*w,2)^2/norm(Dx(Nw), 1); 


%++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
% Validation step.
%++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
% Set initial gamma.
%gammas(:,1)=gamscale*norm(RN*w,2)^2/norm(Dx(N*w),1);

for i=1:num_gammas
    N=N0;
    RN = RN0;
    
    % Set gamma_1.
%     size(RN)
%     size(N)
%     size(w)    
    gammas(:, 1) = gmults(i)*gmax;    
    
    for j=1:K-1
        
        % Initial solutions.
        if i == 1
            sols0.x = w;
            sols0.y = Dx(Nw);
            sols0.z = zeros(p,1);
        else
            sols0.y = DVs(:,j,i-1); % Warm-start with previous DV.
            sols0.x = N'*Dtx(sols0.y); % x = N'*D'*y.
            sols0.z = zeros(p,1);
        end
        quietADMM=1;
        
        %Call ADMM
        
        if isequal(consttype,'ball')
            
            % Call ball-constrained solver.
            [x,y,~,tmpits]=SZVD_ADMM_V2(R,N,RN, D,sols0,gammas(i,j),beta,tol,maxits,quietADMM);
            
            % Normalize y, if necessary.
            DVs(:,j,i) = y/norm(y);
            
        elseif isequal(consttype,'sphere')
            
            % Call spherical solver.
            [x,DVs(:,j,i),~,tmpits]=SZVD_ADMM_S(R,N,RN, D,sols0,gammas(i,j),beta,tol,maxits,quietADMM);
            
        else % ERROR.
            error('Invalid constraint type. Please indicate if using inequality ("ball") or equality ("sphere") constraints.')
        end
          
        its(i)= its(i) + tmpits;
        %Update N and B 
        if j< K-1
            %Project N onto orthogonal complement of Nx 
            x=Dx(N*x);
            x=x/norm(x);
%             fprintf('update N\n')
            N=Nupdate1(N,x);
            RN = R*N;
            [~,sigma,w]=svds(RN, 1, 'largest');
            R=R/sigma;
            % Set gamma.
%             size(RN)
%             size(N)
%             size(w)
            gammas(i,j+1)=gmults(i)*norm(RN*w,2)^2/norm(Dx(N*w),1);
        end
    end
    %Get performance scores on the validation set
    %Call test ZVD to get predictions
    [stats,~,~,~]=predict(DVs(:,:,i), val, classMeans);%%%%%%%%%%%%%
    
    %If gamma gives trivial sol, give it a large penalty
    if (min(stats.l0)<1)
        val_score(i)=(p+1)*size(val,1);
        triv=1;
    else
        if (sum(stats.l0)>sparsity_level*(K-1)*p)
            val_score(i)=sum(stats.l0);
        else
            val_score(i)=stats.mc;
        end
    end
    
    %Update the best gamma so far
    if (val_score(i)<=val_score(best_ind))
        best_ind=i;
    end
    
    %Record sparest nontrivial sol
    %if (min(stats.l0)>3 && stats.l0<min_l0)
    %    l0_ind=1;
    %    l0_x=DVs(:,:,i);
    %    min_10=stats.l0;
    %end
    %Record best misclassificartion error
    %if(stats.mc<=min_mc)
    %    min_ind=i;
    %    mc_x=DVs(:,:,i);
    %    min_mc=stats.mc;
    %end
    %Terminate if a trivial solution has been found
    
    if (quiet==false)
       if (i==1) 
           fprintf('++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n')
           fprintf('i \t\t + gam \t\t\t + score \t\t + mc \t\t\t + l0 \t\t + its \n')
           fprintf('++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n')
       end
       fprintf('%3d \t +  %1.2e \t + %1.2e \t + %1.2e \t + %1.2e \t + %1.2e \n', i, gammas(i), val_score(i),  stats.mc, sum(stats.l0), its(i));
    end
    if(triv==1)
        break;
    end
end


%Export DVs found using validation
val_w=DVs(:,:,best_ind);
gamma=gammas(best_ind,:);
%scaler=gmults(best_ind);
%val_score=val_score(best_ind);

File: R/package/penzda/R/penzdaVAL.R
#' PENZDAVAL - PENalized Zero-variance Discriminant Analysis with Validation.
#'
#' @param Xt training data set (as matrix)
#' @param Yt factor containing labels of training data.
#' @param Xval validation data set (as matrix)
#' @param Yval factor containing labels of validation data.
#' @param D dictionary matrix.
#' @param gmults vector of multipliers defining potential regularization parameter values.
#' @param sparsity_level desired minimum sparsity level for validation scoring.
#' @param tol stopping tolerance for ADMM scheme.
#' @param maxits maximum number of iterations performed by ADMM.
#' @param bta augmented Lagrangian penalty term.
#' @param quiet true suppresses display of intermediate outputs.
#' @param type constraint type: "ball" or "sphere".
#' @return val_w discriminant vectors corresponding to best parameter.
#' @return DVs set of all discriminant vectors calculated.
#' @return gamma optimal regularization parameter choice.
#' @return bestind position of multiplier corresponding to best gamma.
#' @return valscores array of validation scores.
#' @return classMeans set of training data class-means.
#' @export

penzdaVAL <- function(Xt, Yt, Xval, Yval, 
                      D = diag(p), gmults, sparsity_level = 0.25,
                      tol=1e-3, maxits=1000, bta=3, 
                      quiet=FALSE, type="ball")
{
  # Check for valid type.
  stopifnot(exprs = {
    (type=="ball" | type=="sphere")
  } ) 
  
  # Call calcClassMeans.
  cmnsres <- calcClassMeans(Xt = Xt, Yt = Yt)
  
  classMeans <- cmnsres$classMeans
  M <- cmnsres$M
  R <- cmnsres$R
  k <- cmnsres$k
  
  # Find null basis.
  N <- Null(t(M))
  
  # Form factor of objective matrix.
  RN <- R%*%N
  
  if(min(dim(RN)) <= 2){ # k or p <= 2
    # Take SVD (keep first right singular vector)
    SVDres <- svd(x=RN, nu=0, nv = 1)
  }
  else{ 
    # Use Lanczos method to find dominant singular value/right vector.
    SVDres <- rARPACK::svds(A = RN, k =1, nu = 0, nv =1)
  }
  
  # Extract singular value/vector.
  sig <- SVDres$d[1]
  w <- SVDres$v[,1]
  
  Nw <- N %*% w
  
  # Normalize R.
  R <- R/sig
  RN <- RN/sig
  
  # Define d operators.
  if (norm(x= (diag(diag(D)) - D), type='F') < 1e-12){ # D is diagonal.
    
    d <- diag(D) # Extract diagonal of D.
    
    if (norm(x= as.matrix(d - rep(1,p)), type='F') < 1e-12){ # D = I.
      Dx <- function(x){return(x)}
      Dtx <- function(x){return(x)}
    }
    else{ # D is diagonal, but not identity.
      Dx <- function(x){return(d*x)}
      Dtx <- function(x){return(d*x)}
    }
  }
  else{ # treat D as arbitrary matrix.
    Dx <- function(x){return(D%*%x)}
    Dtx <- function(x){return(t(D)%*%x)}
  }
  
  # Initialize scores.
  N0 <- N
  RN0 <- RN
  numgams <- length(gmults)
  val_score <- (p+1)*rep(k-1, numgams)
  best_ind <- 1
  triv <- 0
  quietADMM <- TRUE
  
  # Initialize discriminant vectors.
  DVs <- array(0,c(p,k-1, numgams))
  its <- matrix(0, k-1, numgams)
  
  # Initialize gammas.
  gammas <- matrix(0, numgams, k-1);
  gmax <- norm(RN %*% w,type = 'F')^2/(2*norm(Dx(Nw), type='1')); 
  
  # CALCULATE SETS OF DISCRIMINANT VECTORS AND VALIDATE.
  for (i in 1:numgams){
    
    # Reset N matrix and products.
    N <- N0
    RN <- RN0
    
    # Calculate each discriminant vector corresponding to gmult[i]
    for (j in 1:(k-1)){
      
      # Calculate initial solutions.
      if (i == 1){ # First discriminant vector.
        sols0 = list(x = w, y = Dx(N %*%w), z = rep(0,p))
      } # end if i =1.
      else{ # warm-start with last discriminant vector found for this multiplier.
        sols0 = list(x = t(N) %*% Dtx(as.matrix(DVs[,j, i-1])),
                     y = as.matrix(DVs[,j, i-1]), 
                     z = rep(0,p))
      } # end else
      
      # Choose gamma.
      gammas[i, j] <- gmults[i]*norm(x=RN%*% sols0$x, type = "F")^2/(2*norm(x=sols0$y, type="1"))
      
      # Call ADMM solver.
      ADMMres <- penzdaADMM(R=R, N=N, RN=RN, D=D, 
                            sols0=sols0, gam=gammas[i,j], bta = bta, 
                            tol = tol, maxits = maxits, 
                            type = type, quiet = quiet)
      
      # Save discriminant vector.
      if (norm(ADMMres$y)>1e-12){ # Normalize if not 0 vector.
        DVs[,j,i] <- ADMMres$y/norm(x=ADMMres$y, type="F")
      }
      else{print('Converged to zero')}
      
      
      # Record number of iterations.
      its[j,i] <- ADMMres$its
      
      # Update null-basis.
      if (j < (k-1)){
        # Extract vector to add to rows of W.
        v <- DVs[,j,i]
        
        # Call Nupdate.
        N <- Nupdate(N=N, v = v)
        
        # Update RN and initial solution.
        RN <- R %*% N
        
        if(min(dim(RN)) <= 2){ # k or p <= 2
          # Take SVD (keep first right singular vector)
          SVDres <- svd(x=RN, nu=0, nv = 1)
        }
        else{ 
          # Use Lanczos method to find dominant singular value/right vector.
          SVDres <- rARPACK::svds(A = RN, k =1, nu = 0, nv =1)
        }
        
        # Extract singular value/vector.
        sig <- SVDres$d[1]
        w <- SVDres$v[,1]
        
        # Normalize R.
        R <- R/sig
        RN <- RN/sig
        
        
      } # end if Nupdate
      
    } # End for j.
    
    # UPDATE VALIDATION SCORES FOR ITH SET OF DISCRIMINANT VECTORS.
    # Form penzda object for validation
    obji <- list(DVs = DVs[,,i], k=k, classMeans = classMeans)
    
    # Call predict fxn
    stts <- predict(obji, Xtest = Xval, Ytest = Yval)
    
    if (stts$l0 == 0){ # Found trivial solution 
      val_score[i] <- (p+1)*(k-1) 
      triv <- 1 
      break # Found trivial solution.
    } # End triv solution case.
    else if (stts$l0 > sparsity_level*(k-1)*p){ # Have nontrivial solution, but too dense. 
      val_score[i] <- stts$l0  
    } # End dense case.
    else{ # Sufficiently sparse. Use misclassification rate as score.
      val_score[i] <- stts$mc
    } # End sparse case and validation score update.
    
    # Check if validation score is best so far.
    if (val_score[i] < val_score[best_ind]){ # Min score so far.
      best_ind <- i # Update best_ind.
    }
    
  } # End for i.
  
  # OUTPUT.
  return(list(DVs = DVs[,,best_ind], allDVs = DVs, gam = gammas[i,],
              bestind = best_ind, val_score = val_score,
              its = its, classMeans = classMeans, k=k
  ))
  
} # END PENZDAVAL.